### PR TITLE
Migrate to React@16.3 for Context API

### PR DESCRIPTION
**Pull request by AllenFang**
1. No internal store anymore, more props driven and stateless for low level component
2. Improve data flows
3. Implement table toolkits: like search and csv export

### CODE REVIEW

File: packages/react-bootstrap-table2/src/contexts/index.js [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+            }
+          </SortContext.Consumer>
+        </SortContext.Provider>
+      );
+    }
+
+    renderWithFilterCtx(base, baseProps) {
+      return (
+        rootProps,
+        cellEditprops
```
Comment (Chun-MingChen) [inline packages/react-bootstrap-table2/src/contexts/index.js]: `cellEditprops` => `cellEditProps`

File: packages/react-bootstrap-table2/index.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import BootstrapTable from './src/bootstrap-table';
-import withDataStore from './src/container';
+import withContext from './src/contexts';
 
-export default withDataStore(BootstrapTable);
+export default withContext(BootstrapTable);
```
Comment (Chun-MingChen) [inline packages/react-bootstrap-table2/index.js]: There's extra empty line in `ln. 8`.

**Comment by AllenFang** [comment]
@Chun-MingChen  Just do me a favor, please use `react-create-app` and use npm link to link `react-bootstrap-table-next` to test whether it can backward compatible with < react@16.3.0.  This test will help me to decide `peerDependency` and it's may have a big impact for our user, thanks

**Comment by Chun-MingChen** [comment]
@AllenFang:

Sorry for reply late. I got it and I'll do it instantly.

**Comment by AllenFang** [comment]
Also consider to handle https://github.com/react-bootstrap-table/react-bootstrap-table2/issues/382 . this issue

**Comment by AllenFang** [comment]
https://github.com/react-bootstrap-table/react-bootstrap-table2/issues/244

**Comment by AllenFang** [comment]
#332

**Pull request merged**